src/util/messagepipe.h

```diff
@@ -1,109 +1,94 @@
-        QSharedPointer<TwoWayMessagePipe<SenderMessageType, ReceiverMessageType>> pipe(
-                new TwoWayMessagePipe<SenderMessageType, ReceiverMessageType>(
-                        sender_fifo_size, receiver_fifo_size));
-
-        return {
-                std::make_unique<MessagePipe<SenderMessageType, ReceiverMessageType>>(
-                        pipe->m_receiver_messages,
-                        pipe->m_sender_messages,
-                        new ReferenceHolder<TwoWayMessagePipe<SenderMessageType,
-                                ReceiverMessageType>>(pipe)),
-                std::make_unique<MessagePipe<ReceiverMessageType, SenderMessageType>>(
-                        pipe->m_sender_messages,
-                        pipe->m_receiver_messages,
-                        new ReferenceHolder<TwoWayMessagePipe<SenderMessageType,
-                                ReceiverMessageType>>(pipe))};
-    }
-
-  private:
-    TwoWayMessagePipe(int sender_fifo_size, int receiver_fifo_size)
-            : m_receiver_messages(receiver_fifo_size),
-              m_sender_messages(sender_fifo_size) {
-    }
-
-    // Messages waiting to be delivered to the receiver.
-    rigtorp::SPSCQueue<SenderMessageType> m_receiver_messages;
-    // Messages waiting to be delivered to the sender.
-    rigtorp::SPSCQueue<ReceiverMessageType> m_sender_messages;
-
-    DISALLOW_COPY_AND_ASSIGN(TwoWayMessagePipe);
+/// makeTwoWayMessagePipe is a bare-bones wrapper around the above rigtorp::SPSCQueue class that
+/// facilitates non-blocking two-way communication. To keep terminology clear,
+/// there are two sides to the message pipe, the sender side and the receiver
+/// side. The non-blocking aspect of the underlying rigtorp::SPSCQueue class requires that the
+/// sender methods and target methods each only be called from a single thread
+/// The most common use-case of this class is sending and receiving messages
+/// with the callback thread without the callback thread blocking.
+/// Returns a pair of MessagePipes,
+/// the first is the sender's pipe (sends SenderMessageType and receives
+/// ReceiverMessageType messages) and the second is the receiver's pipe
+/// (sends ReceiverMessageType and receives SenderMessageType messages).
+template<class SenderMessageType, class ReceiverMessageType>
+static std::pair<
+        std::unique_ptr<MessagePipe<SenderMessageType, ReceiverMessageType>>,
+        std::unique_ptr<MessagePipe<ReceiverMessageType, SenderMessageType>>>
+makeTwoWayMessagePipe(
+        int sender_fifo_size,
+        int receiver_fifo_size) {
+    auto sender_pipe = gsl::make_not_null(
+            std::make_shared<rigtorp::SPSCQueue<ReceiverMessageType>>(
+                    sender_fifo_size));
+    auto receiver_pipe =
+            gsl::make_not_null(std::make_shared<rigtorp::SPSCQueue<SenderMessageType>>(
+                    receiver_fifo_size));
+    // TODO: remove unnecessary unique_ptr indirection
```

Comment:
I assume this `TODO` will need to remain for now, could it be updated with the GH issue it relates to?

Reply:
Just made the changes here instead. see newest commit.